src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt

```diff
@@ -671,20 +675,146 @@ data class ImportContext private constructor(
+            scope = RsWithMacrosScope(project, GlobalSearchScope.allScope(project)),
+            pathParsingMode = PathParsingMode.NO_COLONS,
+            attributes = element.stdlibAttributes,
+            namespaceFilter = { true }
+        )
     }
 }
 
-private val RsPath.pathParsingMode: PathParsingMode get() = when (parent) {
-    is RsPathExpr,
-    is RsStructLiteral,
-    is RsPatStruct,
-    is RsPatTupleStruct -> PathParsingMode.COLONS
-    else -> PathParsingMode.NO_COLONS
+/**
+ * @param toImport  Set of unresolved items that can be imported
+ * @param toQualify Set of unresolved items that can't be imported
+ */
+data class TypeReferencesInfo(
+    val toImport: Set<RsQualifiedNamedElement>,
+    val toQualify: Set<RsQualifiedNamedElement>
+)
+
+/**
+ * Traverses type references in `elements` and collects all items that unresolved in current context.
+ */
+fun RsElement.getTypeReferencesInfoFromElements(
+    elements: Collection<RsElement>,
+    subst: Substitution = emptySubstitution
+): TypeReferencesInfo = getTypeReferencesInfo(this, elements, subst, ::collectImportSubjectsFromTypeReferences)
+
+/**
+ * Traverses types in `ty` and collects all items that unresolved in current context.
+ */
+fun RsElement.getTypeReferencesInfoFromTy(
+    ty: Ty,
+    subst: Substitution = emptySubstitution
+): TypeReferencesInfo = getTypeReferencesInfo(this, listOf(ty), subst, ::collectImportSubjectsFromTy)
+
+private fun <T> getTypeReferencesInfo(
+    context: RsElement,
+    elements: Collection<T>,
+    subst: Substitution,
+    collector: (T, Substitution, MutableSet<RsQualifiedNamedElement>) -> Unit
+): TypeReferencesInfo {
+    val result = hashSetOf<RsQualifiedNamedElement>()
+    elements.forEach { collector(it, subst, result) }
+    return processRawImportSubjects(context, result)
+}
+
+private fun collectImportSubjectsFromTypeReferences(
+    context: RsElement,
+    subst: Substitution,
+    result: MutableSet<RsQualifiedNamedElement>
+) {
+    context.accept(object : RsVisitor() {
+        override fun visitTypeReference(reference: RsTypeReference) =
+            collectImportSubjectsFromTy(reference.type, subst, result)
+
+        override fun visitElement(element: RsElement) =
+            element.acceptChildren(this)
+    })
 }
+
+private fun collectImportSubjectsFromTy(
+    ty: Ty,
+    subst: Substitution,
+    result: MutableSet<RsQualifiedNamedElement>
+) {
+    ty.substitute(subst).visitWith(object : TypeVisitor {
+        override fun visitTy(ty: Ty): Boolean {
+            when (ty) {
+                is TyAdt ->
+                    result.add(ty.item)
+                is TyAnon ->
+                    result.addAll(ty.traits.map { it.element })
+                is TyTraitObject ->
+                    result.add(ty.trait.element)
+                is TyProjection -> {
+                    result.add(ty.trait.element)
+                    result.add(ty.target)
+                }
```

Comment:
I think it can be reformatted into 
```kotlin
is TyAdt -> result.add(ty.item)
is TyAnon -> result.addAll(ty.traits.map { it.element })
is TyTraitObject -> result.add(ty.trait.element)
is TyProjection -> {
    result.add(ty.trait.element)
    result.add(ty.target)
}
```

Reply:
And maybe it's worth to use `+=` instead of `add` as more kotlin way.
But it's a matter of taste so I don't insist